### ISSUE TITLE
Reduce span highlighted code in unused_variables lint

### ISSUE THREAD

**Issue by killercup** [rust-lang/rust#50472]
Let's say I wrote this critical piece of code:

```rust
fn main() {
    let mut i = 42;
}
```

rustc gives me the following output:

```
warning: unused variable: `i`
 --> src/main.rs:2:9
  |
2 |     let mut i = 42;
  |         ^^^^^ help: consider using `_i` instead
  |
  = note: #[warn(unused_variables)] on by default

warning: variable does not need to be mutable
 --> src/main.rs:2:9
  |
2 |     let mut i = 42;
  |         ----^
  |         |
  |         help: remove this `mut`
  |
  = note: #[warn(unused_mut)] on by default
```

Which is correct -- except for one little detail. Which is crucial, if you are not just any developer but actually _rustfix_.

You see, the `mut ` in `mut i` is highlighted, but the replacement is just for the binding name, not the mutability. (Why not the mutability? Because the mutability is covered by the next lint we trigger, `unused_mut`.) And _rustfix_ will rightfully complain if we try to replace parts of the code that we have already replaced.

tl;dr We need to change the span in `unused_variables` to only cover the `i`

**Comment by 11Takanori** [comment]
I'd like to take this issue.

**Comment by killercup** [comment]
Gladly! This seems to be place where the lint gets triggered:

https://github.com/rust-lang/rust/blob/6f721f54c6fb1de9cf00eb9d2d050f818c882871/src/librustc/middle/liveness.rs#L1542-L1562

and the `sp` parameter should be the span we care about. I'd probably try and see if it's a binding with a mutability specifier and try to get the sub-span of that. (No idea what the actual API for that is, though, sorry.)

**Comment by killercup** [comment]
@csmoe I'd prefer to somehow get the original span if possible, instead of creating new spans. Sadly, this is not as trivial. A short code dive revealed that the name we use here

https://github.com/rust-lang/rust/blob/6f721f54c6fb1de9cf00eb9d2d050f818c882871/src/librustc/middle/liveness.rs#L1530-L1531

is from

https://github.com/rust-lang/rust/blob/6f721f54c6fb1de9cf00eb9d2d050f818c882871/src/librustc/middle/liveness.rs#L1490

which calls

https://github.com/rust-lang/rust/blob/6f721f54c6fb1de9cf00eb9d2d050f818c882871/src/librustc/middle/liveness.rs#L330-L337

so, our "variable name" might in fact be `<clean-exit>`! Not sure how deep you want to go here, but it might make sense to refactor this a bit.

We should also try to add an applicability marker to this lint (landed in #50204, you can see it in use in #50454).

**Comment by killercup** [comment]
Not necessarily something we need to fix in this exact issue but I just wanted to mention that we should probably also add a macro-check to this lint. For example, this code:

```rust
macro_rules! foo {
    ($x:ident) => {{ let $x = 42; }}
}

fn main() {
    foo!(y);
}
```

tells me this ([playground](https://play.rust-lang.org/?gist=2a0af1d7a401c2f1da18009f19160136&version=stable&mode=debug)):

```
warning: unused variable: `y`
 --> src/main.rs:2:26
  |
2 |     ($x:ident) => {{ let $x = 42; }}
  |                          ^^ help: consider using `_y` instead
...
6 |     foo!(y);
  |     -------- in this macro invocation
  |
  = note: #[warn(unused_variables)] on by default
```

This is somewhat confusing, but should also never be applied by rustfix as suggested.

**Comment by 11Takanori** [comment]
Is this issue difficult for a beginner like me?
It will take me time to understand scope we need to fix.

**Comment by killercup** [comment]
@11Takanori hard to say, I'm neither a regular contributor nor mentor for rustc. Other people did diagnostics adjustments as first contributions to the compiler. Maybe you and @csmoe can figure something out together?

cc @Manishearth and @oli-obk who both know more on this than I do

**Comment by csmoe** [comment]
@11Takanori Are you still here? If you are interested in this issue, this is the instruction:
All the unused_local-checking are initiated by `check_local`:
https://github.com/rust-lang/rust/blob/8ff4b42064b374bb62043f7729f84b6d979c7667/src/librustc/middle/liveness.rs#L1394-L1404
So the issue is caused by the wrong span(`sp`) which represent a `Pat`, but what we want is the REAL name span.

----

https://github.com/rust-lang/rust/blob/8ff4b42064b374bb62043f7729f84b6d979c7667/src/librustc/hir/mod.rs#L1125-L1129
We can find all the info of a variable from a `let` statement(represented by `Local`), as indicated by the comment of struct `Local`, those info might live in `<pat>`. Finally, we get it, all the info are wrapped in the `Spanned` struct:
https://github.com/rust-lang/rust/blob/8ff4b42064b374bb62043f7729f84b6d979c7667/src/librustc/hir/mod.rs#L877
https://github.com/rust-lang/rust/blob/8ff4b42064b374bb62043f7729f84b6d979c7667/src/libsyntax/codemap.rs#L51-L54

----

Now we can derive an approcah from the analysis above:
1. Add a method to get the `span` of the varibale. the [simple_name](https://github.com/rust-lang/rust/blob/8ff4b42064b374bb62043f7729f84b6d979c7667/src/librustc/hir/pat_util.rs#L135-L142) already implements a way to unwrap variable name, so you can create a `simple_span` method, its definition is almost same as `simple_name`.
2. Correct the `sp` in `check_local` with the real `span` we get from step 1.
3. Clean up the remaining trival error.

Use `./x.py check` in between iterations to ensure everything builds at each step. Then, when an error occurs, you can bisect to see why.
It seems your first PR to rustc, so https://forge.rust-lang.org/ and `./configure --help` are highly recommended before coding, for example, `./configure --llvm-root=<path>` and `./x.py test --test-args=ui` can save your time.
Feel free to ping me if you have any problem. :)

----
> We should also try to add an applicability marker to this lint.

We'll step to applicability after `span` lint is done.
